src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs

```diff
@@ -425,32 +431,47 @@ private enum AssignmentQueryKind {
 
 					return AssignmentQueryKind.Hopeless;
 
-				default:
-					var model = m_compilation.GetSemanticModel( assignment.Expression.SyntaxTree );
-					var typeInfo = model.GetTypeInfo( assignment.Expression );
-
-					// Type can be null in the case of an implicit conversion where the
-					// expression alone doesn't have a type. For example:
-					//   int[] foo = { 1, 2, 3 };
-					var typeToCheck = typeInfo.Type ?? typeInfo.ConvertedType;
-
-					if( assignment.Expression is BaseObjectCreationExpressionSyntax _ ) {
-						// When we have a new T() we don't need to worry about the value
-						// being anything other than an instance of T.
-						query = new ImmutabilityQuery(
-							ImmutableTypeKind.Instance,
-							typeToCheck
-						);
-					} else {
-						// In general we need to handle subtypes.
-						query = new ImmutabilityQuery(
-							ImmutableTypeKind.Total,
-							typeToCheck
-						);
+				case SyntaxKind.InvocationExpression:
+					var semanticModel = m_compilation.GetSemanticModel( assignment.Expression.SyntaxTree );
+					var symbol = semanticModel.GetSymbolInfo( assignment.Expression ).Symbol;
+
+					if( symbol is not IMethodSymbol methodSymbol ) {
+						break;
 					}
 
-					return AssignmentQueryKind.ImmutabilityQuery;
+					var symbolName = methodSymbol.ContainingSymbol.ToDisplayString() + "." + methodSymbol.Name;

```

Comment:
Can we map the list of magic things to `ISymbol` and not use string comparison here?

It probably makes sense to go into `ImmutabilityContext` because that is the "is this thing immutable" oracle: it says yes/no based on hard coded lists and the `[Immutable]` attribute... the `ImmutableDefinitionChecker` is the thing that validates the `[Immutable]` attribute was used correctly.

Reply:
Look at how DangerousMethodsAnalyzer loads its `IMethodSymbol`s as a basis. You can simplify it a bit by just having tuples of `(string TypeName, string MethodName)` instead of the dictionary of typename => MethodName[].

And as @j3parker pointed out elsewhere. Filter down to methods with 0 arguments.

https://github.com/Brightspace/D2L.CodeStyle/blob/7e9cde39e549dbc16e916b99b5e3b2e3f1a62f09/src/D2L.CodeStyle.Analyzers/ApiUsage/DangerousMemberUsages/DangerousMemberUsagesAnalyzer.cs#L223-L249

Reply:
Thanks @omsmith .

FYI`ImmutabilityContext` is split into two halves:

* `ImmutabilityContext.Factory.cs`: creates the context by poking around the compilation looking for symbols
* `ImmutabilityContext.cs`: the actual object (e.g. where your `.Contains` would go)

Reply:
Yeah, in slack the only reason I questioned whether ImmutabilityContext was the _right_ suggestion for me to make, is because ImmutabilityContext won't be using it. But not crazy for it to "hold" them and expose afunction to check against it.